plugins/synaptics-mst/fu-synaptics-mst-device.c

```diff
@@ -771,20 +772,147 @@ fu_synaptics_mst_device_panamera_prepare_write (FuSynapticsMstDevice *self, GErr
+						     self->layer,
+						     self->rad);
+	for (guint32 retries_cnt = 0;; retries_cnt++) {
+		guint32 checksum = 0;
+		guint32 flash_checksum = 0;
+
+		if (!fu_synaptics_mst_device_set_flash_sector_erase(self, 0xffff, 0, error))
+			return FALSE;
+		g_debug("Waiting for flash clear to settle");
+		g_usleep(FLASH_SETTLE_TIME);
+
+		for (guint32 i = 0; i < write_loops; i++) {
+			g_autoptr(GError) error_local = NULL;
+			guint8 length = BLOCK_UNIT;
+
+			if (data_to_write < BLOCK_UNIT)
+				length = data_to_write;
+			if (!fu_synaptics_mst_connection_rc_set_command(connection,
+									UPDC_WRITE_TO_EEPROM,
+									length,
+									offset,
+									payload_data + offset,
+									&error_local)) {
+				g_warning("Failed to write flash offset 0x%04x: %s, retrying",
+					  offset,
+					  error_local->message);
+				/* repeat once */
+				if (!fu_synaptics_mst_connection_rc_set_command(
+					connection,
+					UPDC_WRITE_TO_EEPROM,
+					length,
+					offset,
+					payload_data + offset,
+					error)) {
+					g_prefix_error(error,
+						       "can't write flash offset 0x%04x: ",
+						       offset);
+					return FALSE;
+				}
+			}
+			offset += length;
+			data_to_write -= length;
+			fu_device_set_progress_full(FU_DEVICE(self),
+						    (goffset)i * 100,
+						    (goffset)(write_loops - 1) * 100);
+		}
+
+		/* verify CRC */
+		checksum = fu_synaptics_mst_device_get_crc(0, 16, payload_len, payload_data);
+		if (!fu_synaptics_mst_connection_rc_special_get_command(connection,
+									UPDC_CAL_EEPROM_CHECK_CRC16,
+									payload_len,
+									0,
+									NULL,
+									4,
+									(guint8 *)(&flash_checksum),
+									error)) {
+			g_prefix_error(error, "Failed to get flash checksum: ");
+			return FALSE;
+		}
+		if (checksum == flash_checksum)
+			break;
+		g_debug("attempt %u: checksum %x didn't match %x",
+			retries_cnt,
+			flash_checksum,
+			checksum);
+
+		if (retries_cnt > MAX_RETRY_COUNTS) {
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_INVALID_DATA,
+				    "checksum %x mismatched %x",
+				    flash_checksum,
+				    checksum);
+			return FALSE;
+		}
+	}
+
+	if (!fu_synaptics_mst_connection_rc_set_command(connection,
+							UPDC_ACTIVATE_FIRMWARE,
+							0,
+							0,
+							NULL,
+							error)) {
+		g_prefix_error(error, "active firmware failed!!! ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
 static gboolean
 fu_synaptics_mst_device_restart (FuSynapticsMstDevice *self, GError **error)
 {
 	g_autoptr(FuSynapticsMstConnection) connection = NULL;
 	guint8 buf[4] = {0xF5, 0, 0 ,0};
+	gint offset;
 	g_autoptr(GError) error_local = NULL;
 
+	switch (self->family) {
+	case FU_SYNAPTICS_MST_FAMILY_TESLA:
+	case FU_SYNAPTICS_MST_FAMILY_LEAF:
+	case FU_SYNAPTICS_MST_FAMILY_PANAMERA:
+		offset = 0x2000FC;
+		break;
+	case FU_SYNAPTICS_MST_FAMILY_CAYENNE:
+	case FU_SYNAPTICS_MST_FAMILY_SPYDER:
+		offset = 0x2020021C;
+		break;
+	default:
+		offset = 0x2020021C;
```

Comment:
shouldn't this be an error?  Other areas you seem to error out on unknown family